lib/valueflow.cpp

```diff
@@ -1079,6 +1071,38 @@ static void valueFlowArray(TokenList *tokenlist)
     }
 }
 
+static void valueFlowArrayBool(TokenList *tokenlist)
+{
+    for (Token *tok = tokenlist->front(); tok; tok = tok->next()) {
+        if (tok->hasKnownIntValue())
+            continue;
+        const Variable *var = nullptr;
+        bool known = false;
+        std::list<ValueFlow::Value>::const_iterator val =
+            std::find_if(tok->values().begin(), tok->values().end(), std::mem_fn(&ValueFlow::Value::isTokValue));
+        if (val == tok->values().end()) {
+            var = tok->variable();
+            known = true;
+        } else {
+            var = val->tokvalue->variable();
+            known = val->isKnown();
+        }
+        if (!var)
+            continue;
+        if (!var->isArray() || var->isArgument() || var->isStlType())
+            continue;
+        // TODO: Check for function argument
+        if ((tok->valueType() && tok->valueType()->isIntegral()) ||
+            Token::Match(tok->astParent(), "&|&&|%or%|%oror%|%comp%") ||
```

Comment:
If you include `%comp%` you need to check that the other operator is `0`. For instance, I see no pointer to bool conversion here: `arr == ptr`. If you compare something like `arr == true` then the comparison is always false but I think we want a different warning in that case (comparing a pointer and a non-zero integer with sizeof less than sizeof(pointer)).

Reply:
I am not sure about `&` and `|` .. there will be no pointer to bool conversion for these operands as far as I see.

Reply:
I wonder if the result of a parent `!` will be 0?

Reply:
> I am not sure about & and | .. there will be no pointer to bool conversion for these operands as far as I see.

There can be once its propagated(ie `bool b = arr | x`).

Reply:
> I wonder if the result of a parent ! will be 0?

Yes.

Reply:
> No, its always true. Doing if(x) is the same as if(x == true).

I do not agree. I claim that the bool/integer is promoted.

Example code:
```
#include <iostream>

int main()
{
    char buf[10];
    if (buf == true)
        std::cout << "true\n";
    else
        std::cout << "false\n";
    return 0;
}
```

Output:

$ g++ -fpermissive -o 1 1.cpp
1.cpp: In function ‘int main()’:
1.cpp:7:16: warning: ISO C++ forbids comparison between pointer and integer [-fpermissive]

$ ./1
false

In C code you are allowed to compare a integer and pointer but you still get the output "false" for such code.

Reply:
> I think we are both mistaken, the | operator is not allowed to have pointer operands is it?

Yea, you're right, that code doesn't compile.